### PR TITLE
Upgrade to eslint@8

### PR DESCRIPTION
**Pull request by sviat9440**

### CODE REVIEW

**Comment by sviat9440** [comment]
`semantic-release@19.0.2` requires node version >= 14. If we need support node@12, need to roll back the version of `semantic-release`.

**Comment by sviat9440** [comment]
@NullVoxPopuli I suggest removing the automatic detection of printWidth. Because this requires accessing the internal API eslint.

File: .prettierrc.js [hunk 1]
```diff
@@ -5,4 +5,5 @@ module.exports = {
   semi: true,
   singleQuote: true,
   trailingComma: 'es5',
+  endOfLine: 'auto',
```
Comment (NullVoxPopuli) [inline .prettierrc.js:8]: Why this change?

File: README.md [hunk 2]
```diff
@@ -7,8 +7,8 @@
 
 ## ❗️Requirements
 
-- [ESLint](https://eslint.org/) `>= 6`
-- [Node.js](https://nodejs.org/) `>= 10`
+- [ESLint](https://eslint.org/) `>= 8`
```
Comment (NullVoxPopuli) [inline README.md:10]: Dropping support for v7 entirely? How come?

File: lib/rules/decorator-position.js [hunk 3]
```diff
@@ -137,17 +134,14 @@ function decoratorPositionRule(context) {
   }
 
   const userOptions = context.options[0] || {};
-  const filePath = context.getFilename();
-  const { printWidth } = lineLength(userOptions, filePath);
-  const options = normalizeOptions({ ...userOptions, printWidth });
+  const options = normalizeOptions(userOptions);
 
   return {
-    'ClassProperty[decorators.length=1]:exit'(node) {
```
Comment (NullVoxPopuli) [inline lib/rules/decorator-position.js:145]: What advantage does this new selector with siblings have over the previous?

File: lib/rules/decorator-position.js [hunk 4]
```diff
@@ -392,8 +340,7 @@ function lengthAsInline(context, node) {
   // - type annotation (and !)
   // - etc
   return (
-    context.getSourceCode().getText(node).replace(/\s+/, ' ').length +
-    // this is the only way to get indentation?
+    context.getSourceCode().getText(node).replace(/\s+/, ' ').length + // this is the only way to get indentation?
```
Comment (NullVoxPopuli) [inline lib/rules/decorator-position.js:343]: Why this change?

File: lib/rules/decorator-position.js [hunk 5]
```diff
@@ -426,12 +373,9 @@ function decoratorInfo(context, node, decoratorConfig, options) {
 
   const decoratorName = nameOfDecorator(decorator);
   const arity = arityOfDecorator(decorator);
-  const arityMatches =
-    // we don't care what the args are, if they exist
-    decoratorOptions.withArgs === undefined ||
-    // this config requires args, so ensure the decorator has them
-    (decoratorOptions.withArgs === true && arity >= 0) ||
-    // this config requires no args, so ensure the decorator doesn't have them
+  const arityMatches = // we don't care what the args are, if they exist
```
Comment (NullVoxPopuli) [inline lib/rules/decorator-position.js:376]: Why this change?

File: package.json [hunk 6]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "eslint-plugin-decorator-position",
-  "version": "4.0.1",
```
Comment (NullVoxPopuli) [inline package.json:3]: Please don't change the version, here. The release automation handles that

File: tests/helpers/faux-context.js [hunk 7]
```diff
@@ -1,33 +0,0 @@
-'use strict';
```
Comment (NullVoxPopuli) [inline tests/helpers/faux-context.js:1]: Was this file used?

**Review by NullVoxPopuli — COMMENTED** [review]
Thanks for thin PR! I left some comments and questions -- mostly about clarification and why certain changes happened. I believe the diff can be reduced quite a bit.


Did you see this PR? https://github.com/NullVoxPopuli/eslint-plugin-decorator-position/pull/441

It's trying to also support eslint 7, which i think is important for migration purposes. When too many things require breaking change upgrades,, it can frustrate users. My plan is to make this plugin work with v7 and v8, and then users can upgrade their eslint version separately.

Since your main approach to getting v8 support is to delete some of the prettier detection, i may have to try that it my pr -- i've been avoiding it because i really didn't want to lose features. But... Maybe that's just how it's got to be. Idk.

File: .prettierrc.js (hunk 1 above)
Comment (sviat9440) [reply .prettierrc.js:8]: I use Windows as an operating system. In my OS, by default, line breaks are set as CR LF. Prettier highlights the whole file as an error. This setting removes the error if the line breaks are not CRLF. Moreover, for git it does not matter at all.

File: README.md (hunk 2 above)
Comment (sviat9440) [reply README.md:10]: 1. As a result of experiments, I found out that there is no ClassProperty selector in eslint 8 anymore.
2. Changed API for CliEngine.

But we can assign an additional selector that will work in v8, and the old one will work in v7. Also, since the use of CliEngine has been removed, we don't care what API is there.

Today, I will test in v7 version, and maybe I will return its support.

File: lib/rules/decorator-position.js (hunk 3 above)
Comment (sviat9440) [reply lib/rules/decorator-position.js:145]: As a result of experiments, I found out that there is no ClassProperty selector in eslint 8 anymore.

File: lib/rules/decorator-position.js (hunk 5 above)
Comment (sviat9440) [reply lib/rules/decorator-position.js:376]: IDEs default formatting. Rolling back.

File: tests/helpers/faux-context.js (hunk 7 above)
Comment (sviat9440) [reply tests/helpers/faux-context.js:1]: I have not found where it is used and the tests work without it.

File: lib/rules/decorator-position.js (hunk 4 above)
Comment (sviat9440) [reply lib/rules/decorator-position.js]: IDEs default formatting. Rolling back.

**Comment by sviat9440** [comment]
> i've been avoiding it because i really didn't want to lose features.

Maybe just copy the source code of the desired method to CliEngine?

**Comment by sviat9440** [comment]
I found a workaround. Added a postinstall script that creates a file with the export CLIEngine.

**Comment by NullVoxPopuli** [comment]
Awesome! Looks like you had much more success that i did yesterday. I'll pull it down, take a look, and if all is well, merge and release before tomorrow!

File: lib/config/base.js [hunk 8]
```diff
@@ -15,5 +23,30 @@ module.exports = {
...
+  root: true,
+  parser: '@babel/eslint-parser',
+  parserOptions: {
+    ecmaVersion: 2018,
+    sourceType: 'module',
+    ecmaFeatures: {
+      legacyDecorators: true,
+    },
+    babelOptions: {
+      configFile: resolve(__dirname, '../../.babelrc'),
```
Comment (NullVoxPopuli) [inline lib/config/base.js:40]: I don't think we can know what file path the .babelrc will be at -- is there a generic way of specifying this?

File: lib/rules/decorator-position.js [hunk 9]
```diff
@@ -1,20 +1,23 @@
 'use strict';
 
-const { CLIEngine } = require('eslint');
+const { CLIEngine: v7AndEarlier, ESLint: v8Plus } = require('eslint');
 
 const cli = new CLIEngine();
+const ESLint = v7AndEarlier || v8Plus;
```
Comment (NullVoxPopuli) [inline lib/rules/decorator-position.js:6]: I got rid of the postinstall script and did this.

postinstall script won't run on folks' machines who have ignore-scripts enabled

**Comment by NullVoxPopuli** [comment]
I made some changes for eslint 7 / 8 support -- not sure if it's perfect -- but now tests are failing (expected) -- pretty sure it has to do with alternate parsers at in use

**Comment by NullVoxPopuli** [comment]
I wonder if this breaking change should just forbid eslint-parser as well :thinking:

**Comment by sviat9440** [comment]
Do I understand correctly that automatic `printWidth` detection will not work for version 8? Maybe instead of running this script in postinstall, execute it when the rule is first run? He's synchronous.

**Comment by NullVoxPopuli** [comment]
> Do I understand correctly that automatic printWidth detection will not work for version 8? 

correct

> execute it when the rule is first run? He's synchronous.

oh, how would you set that up?

**Comment by NullVoxPopuli** [comment]
I've got some solid progress over here: https://github.com/NullVoxPopuli/eslint-plugin-decorator-position/actions/runs/2645199386

**Comment by sviat9440** [comment]
> oh, how would you set that up?

@NullVoxPopuli look

**Comment by sviat9440** [comment]
@NullVoxPopuli it remains to fix smoke tests

**Comment by sviat9440** [comment]
I won't be able to fix anything anytime soon. Maybe in the second half of the week. You can wait or fix it yourself.

**Comment by sviat9440** [comment]
eslint v6 also work

**Comment by sviat9440** [comment]
@NullVoxPopuli 
1. What about smoke tests? 
2. Are there any edits for me?

**Comment by NullVoxPopuli** [comment]
Thanks! I'll try to take this the rest of the way as i have the smoke tests mostly working on my PR.

The gist is that they've been broken for a bit, and no one knew 🙃
